spec/lib/diaspora_federation/entities/relayable_spec.rb

```diff
@@ -0,0 +1,139 @@
+          expect(argument).to eq(hash[:parent_guid])
+          false
+        else
+          raise "Unexpected callback"
+        end
+      end
+
+      it "doesn't raise anything if correct data were passed" do
+        hash[:author_signature] = Signing.sign_with_key(hash, author_pkey)
+        hash[:parent_author_signature] = Signing.sign_with_key(hash, parent_pkey)
+
+        expect(DiasporaFederation.callbacks).to receive(:trigger).at_least(1).times {|*args| nice_callbacks(*args) }
+        expect { Entities::Relayable.verify_signatures(hash) }.not_to raise_error
+      end
+
+      it "raises when no public key for author was fetched" do
+        expect(DiasporaFederation.callbacks).to receive(:trigger) do |callback|
+          expect(callback).to eq(:get_public_key_by_id)
+          nil
+        end
+
+        expect { Entities::Relayable.verify_signatures(hash) }.to raise_error(
+          Entities::Relayable::SignatureVerificationFailed
+        )
+      end
+
+      it "raises when bad author signature was passed" do
+        hash[:author_signature] = nil
+        expect(DiasporaFederation.callbacks).to receive(:trigger) {|*args| nice_callbacks(*args) }
+
+        expect { Entities::Relayable.verify_signatures(hash) }.to raise_error(
+          Entities::Relayable::SignatureVerificationFailed
+        )
+      end
+
+      it "raises when no public key for parent author was fetched" do
+        hash[:author_signature] = Signing.sign_with_key(hash, author_pkey)
+        expect(DiasporaFederation.callbacks).to receive(:trigger).at_least(1).times do |callback, argument|
+          case callback
+          when :get_public_key_by_post_guid # parent key is required this way
+            nil
+          else
+            nice_callbacks(callback, argument)
+          end
+        end
+
+        expect { Entities::Relayable.verify_signatures(hash) }.to raise_error(
+          Entities::Relayable::SignatureVerificationFailed
+        )
+      end
+
+      it "raises when bad parent author signature was passed" do
+        hash[:author_signature] = Signing.sign_with_key(hash, author_pkey)
+        hash[:parent_author_signature] = nil
+
+        expect(DiasporaFederation.callbacks).to receive(:trigger).at_least(1).times {|*args| nice_callbacks(*args) }
+
+        expect { Entities::Relayable.verify_signatures(hash) }.to raise_error(
+          Entities::Relayable::SignatureVerificationFailed
+        )
+      end
+
+      it "doesn't raise if parent_author_signature isn't set but we're on upstream federation" do
+        hash[:author_signature] = Signing.sign_with_key(hash, author_pkey)
+        hash[:parent_author_signature] = nil
+
+        expect(DiasporaFederation.callbacks).to receive(:trigger).at_least(1).times do |callback, argument|
+          case callback
+          when :post_author_is_local
+            true # the federation is upstream
+          else
+            nice_callbacks(callback, argument)
+          end
+        end
```

Comment:
I don't like this way to mock the callbacks, for several reasons:
1. you know how many times something should be called, you can do `exactly` instead of `at_least`.
2. you define `:post_author_is_local` twice
3. you expect calls that aren't called.

This makes it difficult to see what is expected while reading the test.

I have rewritten this specs and expect exactly what is called for each test. I think this is better readable and easier to understand. And it needs even less lines of code. :)

See: https://gist.githubusercontent.com/SuperTux88/3781cfeddf51ec1e0388/raw/59f35cc1a6cb64e439ed4feba14056910274a45d/relayable_spec.rb